pkg/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,269 @@
+				// just actively refuses connections (e.g. port is blocked)
+				continue
+			}
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (hc *HealthChecker) checkReachabilityItem(r netReachabilityItem) error {
+	target := r.Host + ":" + strconv.Itoa(r.Port)
+	conn, err := hc.toolkit.tcpDialer.Dial("tcp", target)
+	if err != nil {
+		return fmt.Errorf("error trying to connect to %s: %w", target, err)
+	}
+	if conn != nil {
+		if err = conn.Close(); err != nil {
+			return fmt.Errorf("error closing connection: %w", err)
+		}
+	}
+	return nil
+}
+
+func (hc *HealthChecker) checkReachabilityItemWithRetry(r netReachabilityItem) error {
+	var err error
+	for i := 0; i < hc.retries; i++ {
+		err = hc.checkReachabilityItem(r)
+		if err == nil {
+			return nil
+		}
+		if strings.Contains(err.Error(), "refused") || i >= hc.retries-1 {
+			break
+		}
+	}
+	return err
+}
+
+// NetHealthcheckConfig is a struct that holds healtcheck config.
+type NetHealthcheckConfig struct {
+	Interfaces   []string              `yaml:"interfaces,omitempty"`
+	Reachability []netReachabilityItem `yaml:"reachability,omitempty"`
+	Timeout      string                `yaml:"timeout,omitempty"`
+	Retries      int                   `yaml:"retries,omitempty"`
+}
+
+type netReachabilityItem struct {
+	Host string `yaml:"host"`
+	Port int    `yaml:"port"`
+}
+
+// LoadConfig loads healtcheck config from file.
+func LoadConfig(configFile string) (*NetHealthcheckConfig, error) {
+	config := &NetHealthcheckConfig{}
+
+	isMandatory := false
+	if val := os.Getenv(configEnv); val != "" {
+		isMandatory = true
+		configFile = val
+	}
+
+	configExists := false
+
+	read, err := os.ReadFile(filepath.Clean(configFile))
+	if err != nil {
+		if isMandatory {
+			return nil, fmt.Errorf("error opening config file: %w", err)
+		}
+	} else {
+		configExists = true
+	}
+
+	if configExists {
+		err = yaml.Unmarshal(read, &config)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshalling config file: %w", err)
+		}
+	}
+
+	return config, nil
+}
+
+// Toolkit is a helper structure that holds interfaces and functions used by HealthChecker.
+type Toolkit struct {
+	frr        FRRInterface
+	linkByName func(name string) (netlink.Link, error)
+	tcpDialer  TCPDialerInterface
+}
+
+// NewHealthCheckToolkit returns new HealthCheckToolkit.
+func NewHealthCheckToolkit(frr FRRInterface, linkByName func(name string) (netlink.Link, error), tcpDialer TCPDialerInterface) *Toolkit {
+	return &Toolkit{
+		frr:        frr,
+		linkByName: linkByName,
+		tcpDialer:  tcpDialer,
+	}
+}
+
+// NewTCPDialer returns new tcpDialerInterface.
+func NewTCPDialer(dialerTimeout string) TCPDialerInterface {
+	timeout, err := time.ParseDuration(dialerTimeout)
+	if err != nil {
+		seconds, err := strconv.Atoi(dialerTimeout)
+		if err != nil {
+			timeout = time.Second * defaultTCPTimeout
+		} else {
+			timeout = time.Second * time.Duration(seconds)
```

Comment:
How hard to we want to try to "silently do the right thing" here?
Might be worth to at least log a warning that we can't deal with the config the user provided (and will now just assume they probably meant seconds / fallback to the default timeout).

Reply:
> How hard to we want to try to "silently do the right thing" here? 

I'm completely open to suggestions on that. Maybe others would like to give their insights?

Anyway, added logging for now to give user more info on what's going on. That's really good suggestion, btw!